File: frontend/src/types/rag.ts
export interface RagCitation {
  number: number
  chunk_id: string
  document_id: string
  source_title: string
  source_provider: string
  source_url: string
  filename: string
  source_location: string
  excerpt: string
  similarity: number
}

export interface RagAnswer {
  question: string
  answer: string
  insufficient_evidence: boolean
  citations: RagCitation[]
  retrieval_model: string
  generation_model: string
  prompt_version: string
}

export interface DocumentIndexResult {
  document_id: string
  chunk_count: number
  embedding_model: string
}

File: backend/justsupply/domain/rag.py
from dataclasses import dataclass
from uuid import UUID


@dataclass(frozen=True)
class TextChunk:
    index: int
    text: str
    page_number: int | None
    character_start: int
    character_end: int

    @property
    def source_location(self) -> str:
        if self.page_number is not None:
            return f"page {self.page_number}"
        return f"characters {self.character_start + 1}-{self.character_end}"


@dataclass(frozen=True)
class IndexedChunk:
    chunk: TextChunk
    embedding: list[float]


@dataclass(frozen=True)
class RetrievedChunk:
    id: UUID
    document_id: UUID
    text: str
    chunk_index: int
    page_number: int | None
    character_start: int
    character_end: int
    source_title: str
    source_provider: str
    source_url: str
    filename: str
    similarity: float

    @property
    def source_location(self) -> str:
        if self.page_number is not None:
            return f"page {self.page_number}"
        return f"characters {self.character_start + 1}-{self.character_end}"


@dataclass(frozen=True)
class GeneratedAnswer:
    answer: str
    cited_chunk_numbers: list[int]
    insufficient_evidence: bool
    provider_response_id: str | None = None
